force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,9 +140,33 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
-        // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        List<Opportunity> opportunityList = [SELECT Id,Name FROM Opportunity WHERE Name In :oppNames]; // Create a list to hold the existing Opportunities    
+        List<Opportunity> oppsToProcess = new List<Opportunity> (); // Create a list for the Opps to upsert
+        List<Account> accountsByName = [SELECT Id FROM Account WHERE Name = :accountName]; // Query to see if the Account exists
+        Account oppAccount = new Account (); // Create the Account variable
+
+        if (accountsByName.isEmpty() == false) {
+            oppAccount = accountsByName.get(0); // If at least 1 Account was found, assign it to the Account variable
+        } 
+        else {
+            oppAccount.Name = accountName; // Assign the Account Name
+            insert oppAccount; // Make sure that the Account is in the database
+        }
+
+        for (String s : oppNames) { // Loop through each Opportunity name
+            Opportunity newOpp = new Opportunity (Name = s, AccountId = oppAccount.Id, StageName = 'Prospecting', CloseDate = Date.today()); // Create an Opportunity instance with the given name
+            
+            for(Opportunity opp : opportunityList) { // Loop through existing Opportunities
+                if (opp.Name == s) {
+                    newOpp.Id = opp.Id; // If one is found with the correct name already, save the Id in the variable
+                }
```

Comment:
Excellent job only querying for the Opportunity records above with the input names and then using that list to find the opportunity that matches. It is possible in scenarios where you are passed a lot of oppNames or there are several duplicate values, looping through every time could take up resources, but this meets the needs of basic bulkification.